### ISSUE TITLE
Requests occassionally 502

### ISSUE THREAD

**Issue by arian487** [Recidiviz/recidiviz-data#1325]
Its unclear exactly why, the only error we get is:

[error] 33#33: *1126 recv() failed (104: Connection reset by peer) while reading response header from upstream, client: 2002:ab0:8703::, server: , request: "POST /scraper/work/us_ky_allen HTTP/1.1", upstream: "http://172.17.0.1:8080/scraper/work/us_ky_allen", host: "recidiviz-123.appspot.com"

Looking around it doesn't appear there are good reasons for this, we may have to contact Cloud folks directly.

**Comment by arian487** [comment]
This feels relevant.  This happens for jailtracker at least (might happen elsewhere as well) and its possible we are going over the buffer size:

https://stackoverflow.com/questions/22697584/nginx-uwsgi-104-connection-reset-by-peer-while-reading-response-header-from-u

**Comment by arian487** [comment]
SHHing into the instance and logging at nginx logs confirms that the error is on the nginx side.  investigating further.

**Comment by arian487** [comment]
The 502s appear to happen more often when there are more regions running.  When its just one region it happens infrequently.  The problem is definitely related to concurrent requests though because no 502s happen when we run regions in isolation that only have 5qpm vs jailtracker which has 250qpm

**Comment by arian487** [comment]
The error does not seem to be on the gunicorn side, requests don't seem to ever make it to gunicorn layer since I don't see any of the 502s.  Error is likely on the nginx side.

**Comment by arian487** [comment]
More data:

Our gunicorn keep-alive is set to 2 seconds, while the nginx (effectively the loadbalancer) timeout is set to 650.  My working theory is that this causes the connection to no longer be valid on the gunicorn side, while nginx thinks it is.  It almost certainly makes sense to raise the gunicorn keepalive to see if that fixes it.  I can't quite explain why the 502s happen if the QPS is higher and not really when it is lower though.

**Comment by arian487** [comment]
An initial test run of this setting appears to have fixed the problem, I would guess that we may still get 502s but only once every 650 seconds instead of 2 seconds.  I doubt we can avoid that since we only have 1 instance and 2 workers, and having many more instances would be costly.  I'll leave this bug open in case it is not fixed.

**Comment by arian487** [comment]
So the situation was greatly improved, we now see 502s every 650 seconds like I predicted when requests get backed up.  This will completely go away when we deal with the slow execution time

**Comment by jessex** [comment]
Is it actually a 502 every 650 seconds around the clock, or every 650 seconds during scrape sessions, or every 650 seconds when we're at really high load? Still trying to piece together what exactly is going on across the multiple related issues.

**Comment by arian487** [comment]
It is a collection of many 502s (every request that was backed up and didn't get touched by the workers because they were busy) every 650 seconds (but only when we're at high qps).  Note that its every 650 seconds _per worker_ and we have two of them (I plan to up to 3 based on the reading I've done).  So if you look at the logs during high load, you'll see a mass amount of 502s all happen at once when a workers keepalive runs out, and then again (from that same worker) 650 seconds later.  Here is an in depth explanation of what is going on:

We have gunicorn configured with 2 workers, and the 'eventlet' worker class.  Each worker is a process and the load is distributed evenly amongst the two workers.  Eventlet worker type means we use async for I/O operations, this is what we want because most of what happens in /work is completely I/O bound.  However, 'eventlet' needs to monkey patch various libraries to make them async friendly (httplib, file i/o, etc...) and linked in the other issue (https://github.com/Recidiviz/recidiviz-data/issues/1327) I linked to some bugs explaining that eventlet monkey patching may be broken, and even if it is not, gRPC calls are not monkey patched properly and don't do async properly.  What this means is that the workers actually block when issuing gRPC requests instead of continuing on like we should (datastore, pubsub both use gRPC).  So when we have high load, there are a lot of blocking requests which result in a lot of the scraper/work calls being backed up, resulting in them eventually hitting the 650s keepalive resulting in the 502.

I've suggested some potential fixes in the other bug, but at a worst case I believe we can move to using threads instead of async, it would be less memory friendly but it would work fine.
